Protocol version 7 update proposal (#64)
Concordium/concordium-update-proposals: updates/P7.txt

- adds support for querying the module reference and contract name of a smart
  contract instance in smart contracts;
- changes cooldown behavior so that:
   * when validators and delegators reduce their stake, the reduction is
     immediately effective for future stake calculations, and the amount of the
     reduction is looked for a cooldown period;
   * validators and delegators can make further changes to their stake while
     they already have stake in cooldown;
- disallows shielded (or encrypted) transfer transactions (`TransferToEncrypted`, 
  `EncryptedAmountTransfer` and `EncryptedAmountTransferWithMemo`).

                              Background

Smart contract execution costs are reduced in protocol version 7 because of
efficiency improvements in the implementation of smart contract execution.
The revised cost model is based on an execution model where the interpreter
stack is compiled to registers.

The new protocol version also imposes a limit on the nesting depth of
inter-contract calls to 384. This is since each live execution frame can
require up to 32MB of memory, so limiting the number of active execution

> frames in this way safeguards against excessive memory use.

Reviewer: I was actually surprised as well for this, I remember discussions about setting 512 as the limit.
Haven't found anything describing why 384 instead.

> frames in this way safeguards against excessive memory use.

Reviewer 2: My rationale was that this is safer since with 512 you already reach 16GB which is our stated minimum requirement. 

In reality we could set it to 50 and be fine. 384 has a nice property that it is a bit more than the similar bound on Ethereum :)

                                * * *

As of protocol version 7, smart contracts are able to query the module
reference and contract name of a smart contract instance. This can be used
to identify the code that is behind a smart contract instance, for example as
a means of authenticating a smart contract before interacting with it. A
particular motivation for this is to implement a factory pattern, where one
smart contract (the factory) is responsible for initializing instances of
another (the product). On Concordium, one smart contract cannot directly
create instances of another, but it can be used to invoke an initialization
method on another. The additional queries allow the factory to ensure that
the contract instance it is initializing is of the correct type.

                                * * *

> Prior to protocol version 7, when a validator or a delegator reduces or

Author: We typically have a supplemental commentary on the proposal, where it would make sense to do this. The hash of this file is used to reference the protocol update itself, so this document is largely intended to be normative.

no longer contributes to the effective stake of the baker or delegator.
While the cooldown is in effect on an account, no other changes can be made
to the account's stake.

From protocol version 7, when a validator or delegator reduces its stake,
the stake reduction takes effect straight away, but the amount of the
reduction enters a cooldown queue. (Note that "straight away" here does
not mean that is has an instant effect on the lottery and rewards. The
effective stakes of bakers and delegators are sampled in the first block
of the last epoch before each payday, with the sampled stakes being used
from the upcoming payday. Although the change is recorded on the account
immediately, it will be between 1 epoch and 1 payday plus 1 epoch until
the change affects the proof-of-stake lottery used for consensus.)
Stake in the cooldown queue is considered to be inactive stake, and, as
such, does not contribute to the lottery calculation. However, the
inactive stake still cannot be spent, although it is possible to transfer
it back to active stake.

When stake is removed from a validator or delegator, it enters cooldown
at the point that the removed stake is no longer part of the effective

> stake for lottery calculations. Specifically, when the stake is removed,

Reviewer 3: fine with me :)

unstaked balance. Specifically, it is taken first from "pre-pre-cooldown",
then "pre-cooldown", and then from the "cooldown" stakes in decreasing
order of expiry time. An account can also switch freely between being
a validator and being a delegator without its stake entering cooldown.
(Prior to protocol version 7, it was necessary to remove the validator
or delegator on an account, waiting out the cooldown period, before
instating a delegator or validator.) As a consequence of this, from
protocol version 7, no distinction is drawn between the cooldown time
for validators and for delegators: stake entering cooldown expires
after the minimum of both times.

The purpose of these changes is to maximise the flexibility of staking,
while still ensuring that any stake must still be locked for the
cooldown period before it can be spent.

                                * * *

Since the initial protocol version, the Concordium block chain has
supported shielded transfers (also known as encrypted transfers).
This mechanism allows an account to "shield" part of its balance, 
cryptographically encrypting it so that the amount cannot be inspected

> except by the owner of the account (using its secret encryption key).

Reviewer: **Minor detail:** Maybe not the most important detail here, but an authority could decrypt these by convincing a Identity disclosure authority, right?

   br_if            10                              4
   br_table         10 + target arity               7
   return           8 + return arity                2
   call             26 + #arg + #res                6 + #arg + #res
   call_indirect    28 + 2*(#arg + #res)            8 + |_ 1.1*(#arg + #res) _|
   
   drop             2                               0
   select           3                               2

   local.get        3                               0
   local.set        3                               0
   local.tee        3                               0
   global.get       3                               1
   global.set       3                               1

   inn.load*        4                               1
   i32.store        8                               2
   i64.store        10                              2
   i32.store8       5                               2
   i32.store16      8                               2
   i64.store8       7                               2
   i64.store16      9                               2
   i64.store32      10                              2
   memory.size      4                               1

>    memory.grow      10 + 100 * #pages               10 + 100 * #pages

Reviewer: The optimizations introduced are essentially reducing the need to push and pop information on the stack (hence the local.get/set is now 0), however the cost of this instruction is dominated by the need to allocate memory, so should be pretty much the same.

But in the end, the costs are determined by the benchmarks [here](https://github.com/Concordium/concordium-base/blob/main/smart-contracts/wasm-chain-integration/benches/wasm.rs)

>    memory.grow      10 + 100 * #pages               10 + 100 * #pages

Reviewer 4: Would it be worth mentioning this in the proposal?

   inn.ctz          3                               1
   inn.popcnt       3                               1
   inn.eq           4                               1
   inn.ne           4                               1
   inn.lt*          4                               1
   inn.gt*          4                               1
   inn.le*          4                               1
   inn.ge*          4                               1
   inn.add          4                               1
   inn.sub          4                               1
   inn.mul          5                               2
   inn.div*         5                               2
   inn.rem*         5                               2
   inn.and          4                               1
   inn.or           4                               1
   inn.xor          4                               1
   inn.shl          4                               1
   inn.shr*         4                               1
   inn.rotl         4                               1
   inn.rotr         4                               1
   i32.wrap_i64     3                               1
   inn.extend*      3                               1

> 

Author: I think this could be done in the commentary.

   (Here, target arity and return arity are the arity of the target
   label and the current activation frame respectively; #arg and
   #res are the numbers of arguments and results of the function
   being called; and #pages is the number of memory pages requested
   by memory.grow.)

   The cost (in Energy) of looking up a contract module is reduced from
   |_ size / 50 _| to |_ size / 500 _|, where size is the size of the
   module's WebAssembly source in bytes, not including custom sections.
   This cost is incurred on contract initialization, update, upgrading and
   inter-contract calls.
   
2. The maximum nesting depth of V1 contract calls is limited to 384.

3. In V1 smart contracts, the `invoke` operations allows for two

>    additional operations:

Author: The background section is for the motivation, and here we spell out the specific changes in detail.

                                * * *

As of protocol version 7, smart contracts are able to query the module
reference and contract name of a smart contract instance. This can be used
to identify the code that is behind a smart contract instance, for example as
a means of authenticating a smart contract before interacting with it. A
particular motivation for this is to implement a factory pattern, where one
smart contract (the factory) is responsible for initializing instances of
another (the product). On Concordium, one smart contract cannot directly
create instances of another, but it can be used to invoke an initialization
method on another. The additional queries allow the factory to ensure that
the contract instance it is initializing is of the correct type.

                                * * *

Prior to protocol version 7, when a validator or a delegator reduces or
removes its stake, the amount of the stake reduction is held in a cooldown
state. During the cooldown period, the stake cannot be spent, and remains
part of the effective stake (used for lottery and reward calculation). Once
the cooldown period has elapsed, the stake is available to be spent, and

> no longer contributes to the effective stake of the baker or delegator.

Reviewer 5: ```suggestion
no longer contributes to the effective stake of the validator or delegator.
```

While the cooldown is in effect on an account, no other changes can be made
to the account's stake.

From protocol version 7, when a validator or delegator reduces its stake,
the stake reduction takes effect straight away, but the amount of the
reduction enters a cooldown queue. (Note that "straight away" here does
not mean that it has an instant effect on the lottery and rewards. The

> effective stakes of bakers and delegators are sampled in the first block

Reviewer 5: ```suggestion
effective stakes of validators and delegators are sampled in the first block
```

   
2. The maximum nesting depth of V1 contract calls is limited to 384.

3. In V1 smart contracts, the `invoke` host function allows for two
   additional operations:

   - Query smart contract module reference:
      * Instruction tag: 7
      * Payload:
        - contract index (8 bytes, little endian)
        - contract subindex (8 bytes, little endian)
      * Return codes:
        - 0: success
           the return value is the module reference of the specified
           contract instance
        - 3: missing contract
           there is no contract instance with the given address
      * Cost: 200 Energy
   - Query smart contract name:
      * Instruction tag: 8
      * Payload:
        - contract index (8 bytes, little endian)
        - contract subindex (8 bytes, little endian)
      * Return codes:
        - 0: success
           the return value is the name of the `init` method used to
           create the contract instance (including the "init_" prefix)
        - 3: missing contract
           there is no contract instance with the given address
      * Cost: 200 Energy

4. The following changes apply to staking:
   - The stake of an account is divided into

>      * Active stake: this determines the baker or validator's

Reviewer 5: ```suggestion
     * Active stake: this determines the validator or delegator's
```
